File: bot-discord/backend-py/app/services/menu_3_quota.py
from ..adapters import system, system_mutations
from ..utils.response import error_response, ok_response
from ..utils.validators import (
    require_bool_param,
    require_param,
    require_positive_float_param,
    require_positive_int_param,
    require_protocol,
    require_username,
)


def handle(action: str, params: dict, settings) -> dict:
    if action == "summary":
        title, msg = system.op_quota_summary()
        return ok_response(title, msg)

    if action == "detail":
        ok_p, proto_or_err = require_param(params, "proto", "Quota Detail")
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_param(params, "username", "Quota Detail")
        if not ok_u:
            return user_or_err
        title, msg = system.op_quota_detail(proto_or_err.lower(), user_or_err)
        return ok_response(title, msg)

    if action == "set_quota_limit":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Set Limit"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_q, quota_or_err = require_positive_float_param(params, "quota_gb", title)
        if not ok_q:
            return quota_or_err
        ok_m, t, m = system_mutations.op_quota_set_limit(proto_or_err, user_or_err, float(quota_or_err))
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_set_limit_failed", t, m)

    if action == "reset_quota_used":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Reset Used"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_m, t, m = system_mutations.op_quota_reset_used(proto_or_err, user_or_err)
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_reset_used_failed", t, m)

    if action == "manual_block":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Manual Block"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_e, enabled_or_err = require_bool_param(params, "enabled", title)
        if not ok_e:
            return enabled_or_err
        ok_m, t, m = system_mutations.op_quota_manual_block(proto_or_err, user_or_err, bool(enabled_or_err))
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_manual_block_failed", t, m)

    if action == "ip_limit_enable":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - IP Limit"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_e, enabled_or_err = require_bool_param(params, "enabled", title)
        if not ok_e:
            return enabled_or_err
        ok_m, t, m = system_mutations.op_quota_ip_limit_enable(proto_or_err, user_or_err, bool(enabled_or_err))
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_ip_limit_toggle_failed", t, m)

    if action == "set_ip_limit":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Set IP Limit"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_l, lim_or_err = require_positive_int_param(params, "ip_limit", title, minimum=1)
        if not ok_l:
            return lim_or_err
        ok_m, t, m = system_mutations.op_quota_set_ip_limit(proto_or_err, user_or_err, int(lim_or_err))
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_set_ip_limit_failed", t, m)

    if action == "unlock_ip_lock":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Unlock IP Lock"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_m, t, m = system_mutations.op_quota_unlock_ip_lock(proto_or_err, user_or_err)
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_unlock_ip_failed", t, m)

    if action == "set_speed_download":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Speed Download"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_v, val_or_err = require_positive_float_param(params, "speed_down_mbit", title)
        if not ok_v:
            return val_or_err
        ok_m, t, m = system_mutations.op_quota_set_speed_down(proto_or_err, user_or_err, float(val_or_err))
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_speed_down_failed", t, m)

    if action == "set_speed_upload":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Speed Upload"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_v, val_or_err = require_positive_float_param(params, "speed_up_mbit", title)
        if not ok_v:
            return val_or_err
        ok_m, t, m = system_mutations.op_quota_set_speed_up(proto_or_err, user_or_err, float(val_or_err))
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_speed_up_failed", t, m)

    if action == "speed_limit":
        if not settings.enable_dangerous_actions:
            return error_response("forbidden", "Quota & Access Control", "Dangerous actions dinonaktifkan via env.")
        title = "Quota - Speed Limit"
        ok_p, proto_or_err = require_protocol(params, title)
        if not ok_p:
            return proto_or_err
        ok_u, user_or_err = require_username(params, title)
        if not ok_u:
            return user_or_err
        ok_e, enabled_or_err = require_bool_param(params, "enabled", title)
        if not ok_e:
            return enabled_or_err
        ok_m, t, m = system_mutations.op_quota_speed_limit(proto_or_err, user_or_err, bool(enabled_or_err))
        if ok_m:
            return ok_response(t, m)
        return error_response("quota_speed_toggle_failed", t, m)

    return error_response("unknown_action", "Quota & Access Control", f"Action tidak dikenal: {action}")

File: bot-discord/gateway-ts/src/views/menu_3_quota.ts
import type { MenuDefinition } from "./types";

export const menu3: MenuDefinition = {
  id: "3",
  label: "Quota & Access Control",
  description: "Ringkasan, detail, dan kontrol quota/manual block/IP limit/speed limit per user.",
  actions: [
    { id: "summary", label: "Quota Summary", mode: "direct", style: "primary" },
    {
      id: "detail",
      label: "Quota Detail",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "Quota Detail",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "contoh: alice" },
        ],
      },
    },
    {
      id: "set_quota_limit",
      label: "Set Quota Limit",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "Set Quota Limit",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
          { id: "quota_gb", label: "Quota GB", style: "short", required: true, placeholder: "100" },
        ],
      },
    },
    {
      id: "reset_quota_used",
      label: "Reset Quota Used",
      mode: "modal",
      style: "danger",
      confirm: true,
      modal: {
        title: "Reset Quota Used",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
        ],
      },
    },
    {
      id: "manual_block",
      label: "Manual Block ON/OFF",
      mode: "modal",
      style: "danger",
      modal: {
        title: "Manual Block",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
          { id: "enabled", label: "Enabled", style: "short", required: true, placeholder: "on / off" },
        ],
      },
    },
    {
      id: "ip_limit_enable",
      label: "IP Limit ON/OFF",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "IP Limit Enable/Disable",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
          { id: "enabled", label: "Enabled", style: "short", required: true, placeholder: "on / off" },
        ],
      },
    },
    {
      id: "set_ip_limit",
      label: "Set IP Limit",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "Set IP Limit",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
          { id: "ip_limit", label: "IP Limit", style: "short", required: true, placeholder: "2" },
        ],
      },
    },
    {
      id: "unlock_ip_lock",
      label: "Unlock IP Lock",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "Unlock IP Lock",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
        ],
      },
    },
    {
      id: "set_speed_download",
      label: "Set Speed Download",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "Set Speed Download",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
          { id: "speed_down_mbit", label: "Speed Down Mbps", style: "short", required: true, placeholder: "20" },
        ],
      },
    },
    {
      id: "set_speed_upload",
      label: "Set Speed Upload",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "Set Speed Upload",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
          { id: "speed_up_mbit", label: "Speed Up Mbps", style: "short", required: true, placeholder: "10" },
        ],
      },
    },
    {
      id: "speed_limit",
      label: "Speed Limit ON/OFF",
      mode: "modal",
      style: "secondary",
      modal: {
        title: "Speed Limit Enable/Disable",
        fields: [
          { id: "proto", label: "Protocol", style: "short", required: true, placeholder: "vless/vmess/trojan" },
          { id: "username", label: "Username", style: "short", required: true, placeholder: "alice" },
          { id: "enabled", label: "Enabled", style: "short", required: true, placeholder: "on / off" },
        ],
      },
    },
  ],
};
